### ISSUE TITLE
Updating swarm services from private repository causes all service tasks to fail if token is invalid

### ISSUE THREAD

**Issue by darklow** [docker/docker#31819]
I am using private docker repository on Amazon Container Registry. For nginx service with 3 replicas I accidentally forgot to add `--with-registry-auth` while executing `docker service update` and ended up with all service tasks to fail. 

I run following command:
```
$ docker service update --image 333435094895.dkr.ecr.us-east-1.amazonaws.com/swarm/os:latest 30sec_nginx

Output:
unable to pin image 333435094895.dkr.ecr.us-east-1.amazonaws.com/swarm/os:latest to digest: denied: Your Authorization Token has expired. Please run 'aws ecr get-login' to fetch a new one.
30sec_nginx
```

I though I was safe, as I have 3 replicas for 30sec_nginx service, but while looking at what went went wrong and logging in to refresh token, I noticed swarm kept updating all tasks and eventually I ended up with downtime. I am using defaults `--update-failure-action=pause` and `--update-parallelism=1`

**Question:** In such case scenario as mine is it expected behaviour that swarm keeps updating other service tasks and doesn't treat image update fail as error and pause updating rest of the service tasks? Or is it a bug? Thanks.

![image](https://cloud.githubusercontent.com/assets/445304/23900967/dfd9fdc0-08c3-11e7-8f07-b2758f5b6584.png)

**Output of `docker version`:**

```
Client:
 Version:      17.03.0-ce
 API version:  1.26
 Go version:   go1.7.5
 Git commit:   60ccb22
 Built:        Thu Feb 23 10:40:59 2017
 OS/Arch:      darwin/amd64

Server:
 Version:      1.13.1
 API version:  1.26 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   092cba3
 Built:        Wed Feb  8 06:50:14 2017
 OS/Arch:      linux/amd64
 Experimental: true
```

<details><summary>service inspect output</summary><p>

```
[
    {
        "ID": "2ptp2nmtjsbzkdhaf483okosl",
        "Version": {
            "Index": 21585
        },
        "CreatedAt": "2017-03-03T15:05:54.941073617Z",
        "UpdatedAt": "2017-03-14T18:15:55.673751485Z",
        "Spec": {
            "Name": "30sec_nginx",
            "Labels": {
                "com.docker.stack.namespace": "30sec"
            },
            "TaskTemplate": {
                "ContainerSpec": {
                    "Image": "333435094895.dkr.ecr.us-east-1.amazonaws.com/swarm/nginx:latest@sha256:bc988860ba89bfc454a6a1bbe4937bfea9bcaa734082c8a7274cbfac4a0d0dd8",
                    "Labels": {
                        "com.docker.stack.namespace": "30sec"
                    }
                },
                "Resources": {},
                "Placement": {},
                "ForceUpdate": 0
            },
            "Mode": {
                "Replicated": {
                    "Replicas": 3
                }
            },
            "UpdateConfig": {
                "Parallelism": 1,
                "Delay": 2000000000,
                "FailureAction": "pause",
                "Monitor": 2000000000,
                "MaxFailureRatio": 0
            },
            "Networks": [
                {
                    "Target": "v3tif34390rrruyyht1iu8wf0",
                    "Aliases": [
                        "nginx"
                    ]
                }
            ],
            "EndpointSpec": {
                "Mode": "vip",
                "Ports": [
                    {
                        "Protocol": "tcp",
                        "TargetPort": 80,
                        "PublishedPort": 80,
                        "PublishMode": "ingress"
                    },
                    {
                        "Protocol": "tcp",
                        "TargetPort": 81,
                        "PublishedPort": 81,
                        "PublishMode": "ingress"
                    }
                ]
            }
        },
        "PreviousSpec": {
            "Name": "30sec_nginx",
            "Labels": {
                "com.docker.stack.namespace": "30sec"
            },
            "TaskTemplate": {
                "ContainerSpec": {
                    "Image": "333435094895.dkr.ecr.us-east-1.amazonaws.com/swarm/nginx:latest@sha256:bc988860ba89bfc454a6a1bbe4937bfea9bcaa734082c8a7274cbfac4a0d0dd8",
                    "Labels": {
                        "com.docker.stack.namespace": "30sec"
                    }
                },
                "Resources": {},
                "Placement": {},
                "ForceUpdate": 0
            },
            "Mode": {
                "Replicated": {
                    "Replicas": 3
                }
            },
            "UpdateConfig": {
                "Parallelism": 1,
                "Delay": 2000000000,
                "FailureAction": "pause",
                "Monitor": 2000000000,
                "MaxFailureRatio": 0
            },
            "Networks": [
                {
                    "Target": "v3tif34390rrruyyht1iu8wf0",
                    "Aliases": [
                        "nginx"
                    ]
                }
            ],
            "EndpointSpec": {
                "Mode": "vip",
                "Ports": [
                    {
                        "Protocol": "tcp",
                        "TargetPort": 80,
                        "PublishedPort": 80,
                        "PublishMode": "ingress"
                    },
                    {
                        "Protocol": "tcp",
                        "TargetPort": 81,
                        "PublishedPort": 81,
                        "PublishMode": "ingress"
                    }
                ]
            }
        },
        "Endpoint": {
            "Spec": {
                "Mode": "vip",
                "Ports": [
                    {
                        "Protocol": "tcp",
                        "TargetPort": 80,
                        "PublishedPort": 80,
                        "PublishMode": "ingress"
                    },
                    {
                        "Protocol": "tcp",
                        "TargetPort": 81,
                        "PublishedPort": 81,
                        "PublishMode": "ingress"
                    }
                ]
            },
            "Ports": [
                {
                    "Protocol": "tcp",
                    "TargetPort": 80,
                    "PublishedPort": 80,
                    "PublishMode": "ingress"
                },
                {
                    "Protocol": "tcp",
                    "TargetPort": 81,
                    "PublishedPort": 81,
                    "PublishMode": "ingress"
                }
            ],
            "VirtualIPs": [
                {
                    "NetworkID": "zq5d57lgayaa5zatop974clhm",
                    "Addr": "10.255.0.8/16"
                },
                {
                    "NetworkID": "v3tif34390rrruyyht1iu8wf0",
                    "Addr": "10.0.0.9/24"
                }
            ]
        },
        "UpdateStatus": {
            "State": "completed",
            "StartedAt": "2017-03-14T18:15:11.575526073Z",
            "CompletedAt": "2017-03-14T18:15:55.673738316Z",
            "Message": "update completed"
        }
    }
]

```

</p></details>

**Comment by aluzzardi** [comment]
@darklow Could you include a `docker service inspect <service name>`?

Thanks

**Comment by darklow** [comment]
@aluzzardi updated issue with service inspect

**Comment by aaronlehmann** [comment]
Hi @darklow - I would expect the service to pause after the first pull fails. I tried to reproduce this with the following commands:

```
docker service create --name test --replicas 5 busybox top
docker service update --image doesnotexist test
```

I found that the update paused after the first task was rejected due to failure to pull the image:

```
        "UpdateStatus": {
            "State": "paused",
            "StartedAt": "2017-03-14T22:45:35.330016736Z",
            "CompletedAt": "1970-01-01T00:00:00Z",
            "Message": "update paused due to failure or early termination of task mcamdc8y7rcpie04iloax876b"
        }
```

I noticed that your service has `--update-monitor` set to 2 seconds. I thought that might be related. It means that only failures within the first two seconds of starting the task will cause the update to pause. But I tried testing with `--update-delay 2s --update-monitor 2s` and I still couldn't reproduce the problem. Also, unless I'm missing something, it looks like a failure to pull would be treated as satisfying any `--update-monitor` specification - the failure is considered immediate, as the container never even reaches a running state.

Another thought is that if something causes the tasks to restart on their own (say, if they crash or exit, or if the node where they're running goes down), they would use the new service configuration and therefore not be able to start, even though the initial update was paused.

It would be really helpful if you manage to find a way to reproduce this using a public registry.

**Comment by darklow** [comment]
@aaronlehmann Thanks for looking into. I tried your commands (with and without --update-monitor) and couldn't replicate on public repositories either. Only first replica would fail and others kept running. Can't it be something related to private repositories only? In your test you changed image path to non-existing, while in my case I used same image path for update, but once without "--with-registry-auth" and then all replicas failed. I could try to replicate the test on private repository tomorrow.
